Meta: Don't tell to run script as root when that's not the actual error

demostanis opened the pull request:
Before, and that was the cause of many confusion in #build-problems
on Discord, the Meta/build-image-qemu.sh script would error out with
the message "this script needs to run as root" while the actual error
was unrelated.
![image](https://user-images.githubusercontent.com/40673815/188267406-f4f6e8d8-3741-442b-8759-b716dfe1749b.png)
![image](https://user-images.githubusercontent.com/40673815/188267409-175fa284-49ee-4a32-b088-39e880736820.png)
![image](https://user-images.githubusercontent.com/40673815/188267413-c3f0acfb-e29c-48a0-bf26-d1f1055d76a5.png)

timschumi commented on Meta/build-image-qemu.sh:
```diff
@@ -31,7 +31,17 @@ if [ "$(id -u)" != 0 ]; then
         USE_FUSE2FS=1
     else
-        ${SUDO} -E -- "$0" "$@" || die "this script needs to run as root"
+        ${SUDO} -E -- sh -c "$0"' '"$*"'|| exit 42' || \
```
The quotation marks make this a bit hard to read. Don't we have some other way to construct the same string?

timschumi replied:
That said, I'm pretty sure that what we are doing here messes up the string splitting anyways, so we might as well do the following:

```suggestion
        ${SUDO} -E -- sh -c "$0 $* || exit 42" || \
```

demostanis replied:
is $* even needed here? looks like the script doesn't take any arguments.
$0 not having quotes might eventually error out if someone tries to execute the script manually with its full path... which is rare, but nevertheless a case to be taken into account...

timschumi replied:
I don't think there is any way to preserve spaces in arguments when using `sh -c`, unless one wants to go through the effort of manually wrapping each part in `'` or `"`.

The reason why `$*` is there at all is because this is a pretty generic code snippet for "how do I rerun a script with root permissions". But in the end, I don't think it hurts any, we'd have to be concerned with spaces either way because of `$0`.

demostanis replied:
quoted $0 in the latest changes, but not $* as im not sure that's really required

timschumi reviewed (changes requested):
Similar code exists in all of the `build-image-` scripts (and `build-native-partition.sh`), so we might as well change all of them in one go.

demostanis commented:
stale bot strikes again

timschumi commented:
I just got around to testing this, and noticed that this doesn't print the intended error message if `sudo` fails. I'm pretty sure that `exec` is still breaking this (as it just completely replaces the whole process, including the pending commands), and the only reason that we were getting the `"This script needs to run as root"` message before is because `build-image-qemu` was missing that `exec`.

To fix, I'd suggest applying this across all of the affected scripts:

```diff
diff --git a/Meta/build-image-qemu.sh b/Meta/build-image-qemu.sh
index f1e2f7ff76..a3232ef572 100755
--- a/Meta/build-image-qemu.sh
+++ b/Meta/build-image-qemu.sh
@@ -31,7 +31,7 @@ if [ "$(id -u)" != 0 ]; then
     if [ -x "$FUSE2FS_PATH" ] && $FUSE2FS_PATH --help 2>&1 |grep fakeroot > /dev/null; then
         USE_FUSE2FS=1
     else
-        exec ${SUDO} -E -- sh -c "\"$0\" $* || exit 42" || \
+        ${SUDO} -E -- sh -c "\"$0\" $* || exit 42" || \
         {
             case $? in
                 1)
@@ -40,6 +40,9 @@ if [ "$(id -u)" != 0 ]; then
                 42)
                     exit 1
                     ;;
+                *)
+                    exit 0
+                    ;;
             esac
         }
     fi
```

demostanis commented:
Thanks, I should have been more rigorous in my testing...

timschumi commented:
This suspiciously fails some of the tests that require the disk image scripts, but I don't see how that is caused by your change. Would you mind to rebase it onto the latest master branch to make sure that it was/wasn't a fluke?

demostanis commented:
that error message shows up when mount isn't run as root.
we weren't actually exiting when sudo worked probably, because of `||` (which executes the part inside the `{}` *only* if sudo failed)
thus, i removed the `||` so that it always executes, and added `set +e` so that we can handle sudo exit status even if it fails (instead of exiting the script)

the fact this PR took me so many attempts makes me realize my shell scripting-fu isn't as good as i thought...

demostanis commented:
wrong tag, the PR has been fixed, it needs review now

demostanis commented:
@kleinesfilmroellchen

timschumi commented on Meta/build-image-qemu.sh:
```diff
@@ -31,8 +31,22 @@ if [ "$(id -u)" != 0 ]; then
     else
-        ${SUDO} -E -- "$0" "$@" || die "this script needs to run as root"
-        exit 0
+        # we don't want to exit on error here,
+        # to be able to handle sudo exit status
+        set +e
+
+        ${SUDO} -E -- sh -c "\"$0\" $* || exit 42"
```
The comment doesn't follow the usual comment style, but I'd argue that it's superfluous anyways, so...

```suggestion
        set +e
        ${SUDO} -E -- sh -c "\"$0\" $* || exit 42"
```

demostanis replied:
removed the comment

timschumi reviewed (changes requested):
Last one, after this the PR gets a green check mark from me. :^)
